cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java

```diff
@@ -358,23 +359,25 @@ private InetAddress getLocalHost() {
 
   private TwillApplication createTwillApplication(final Map<String, Integer> instanceCountMap) {
     try {
-      return new MasterTwillApplication(cConf, getSavedCConf(), getSavedHConf(), isExploreEnabled, instanceCountMap);
+      return new MasterTwillApplication(cConf, getSavedCConf(), getSavedHConf(), exploreEnabled, instanceCountMap);
     } catch (Exception e) {
       throw  Throwables.propagate(e);
     }
   }
 
   private void scheduleSecureStoreUpdate(TwillRunner twillRunner) {
-    if (User.isHBaseSecurityEnabled(hConf)) {
+    boolean secureHive =
+      cConf.getBoolean(Constants.Explore.EXPLORE_ENABLED) && UserGroupInformation.isSecurityEnabled();
+    if (User.isHBaseSecurityEnabled(hConf) || secureHive) {
```

Comment:
I think it is sufficient if the condition is 
`User.isHBaseSecurityEnabled(hConf) || UserGroupInformation.isSecurityEnabled()`